ci: add .github/workflows/tests.yml

dontlaugh opened the pull request:
This runs against rakudo-star:latest for now. We'll need to find a way to test against a list of recent versions to detect regressions.

dontlaugh committed:
ci: add .github/workflows/tests.yml

This runs against rakudo-star:latest for now. We'll need to find
a way to test against a list of recent versions to detect regressions.

dontlaugh commented:
@finanalyst This will simply run bin/test-script.sh, which is a test of `zef install`.

finanalyst commented:
@dontlaugh I haven't looked at this since last we met. I've had some deeper problems with the distribution. I think I have sorted them, and I will look at this soon.

dontlaugh committed:
Add if directive that skips the whole workflow

dontlaugh commented:
I've pushed another commit that skips the workflow. This will let you merge it now and turn it on when you're ready.

`if: false` is what i used but [other boolean expressions are possible](https://docs.github.com/en/actions/writing-workflows/choosing-when-your-workflow-runs/using-conditions-to-control-job-execution)

finanalyst commented:
@dontlaugh I'm not sure how to continue with this approach.
If you look at the Dockerfile in the root of the repo, it now creates an image from the very latest rakudo, then adds extra libraries, then compiles this repo.

If the image is run with a terminal, then 
```
cd /opt/rakuast-rakudoc-render
prove6 xt/*
```
runs through all the tests in the repo.

dontlaugh committed:
turn tests back on

dontlaugh committed:
Add prove6 invocation to bin/test-script.sh

This is an unoptimized "do everything" script, but it is enough to
get CI started.

dontlaugh committed:
Make our script only run unit tests

dontlaugh committed:
Build the integration test container, too

dontlaugh committed:
explicitly set ubuntu-24.04 on gha

dontlaugh commented:
I've fixed up this branch. In my opinion, it's ready to merge, but there are some optimizations we ought to consider.

We have two test stanzas in the **.github/workflows/test.yml** file now:

1. `unit_tests` - This is similar to what we had originally. It runs **bin/ci-unit-tests.sh** inside the open source rakudo image. It runs fast.
2. `integration_tests` - This is new. Using the Dockerfile in the project root, it builds a container with a dummy name "project-test-image". The name doesn't matter because we're not pushing the container image to a registry (yet). Then it invokes the container with `sh -c 'prove6 xt/*'` as the command. It also sets up the working directory. It takes longer, obviously.

Everything is green and A-OK on the Raku org's repo https://github.com/Raku/rakuast-rakudoc-render/actions/runs/12565912268/job/35030717118

After you merge this, tests will start running in this repo as well.

dontlaugh commented:
I have a few ideas on making the tests more efficient, but it can wait for a follow up PR.

If you can keep the tests green, I can make them fast.

finanalyst commented:
@dontlaugh This is great. Thank you. However... there are some current problems:
- rakuast-rakudoc-render relies on a change to the rakuast compiler that is currently ahead of 12.2024 (three markup letters were added to a set, and I had to special case them). The changes will be incorporated into 01.2025
- when I created a docker image that passes the tests, I had to recompile rakudo because of the above.
- In order to pass the tests, the environment has to have 'sass' and 'dot' included. There are debian packages for these: dart-sass and graphviz
- tl;dr use `bin/force-compile` and not `zef test`. There are some problems using zef with modules that use Rakudo AST. I don't quite know why. I've mentioned it to Nick and Liz, but its not yet urgent to push them on it. There is a simple workaround - the modules in my distribution have to be installed in a specific order (basically the modules that `use experimental :rakuast` have to be compiled first - I think). And the `.precompiled` directory needs to be nuked. Then the test in `t/` needs to be run. A each `use-ok` is run, it compiles that module. These are both done by `bin/force-compile`, but the two lines there are simple, so could be included in the workflow

I would dearly love to have automatic tests but getting the tests to work locally has been such a pain, I havent had time to set it up automatically. So your help here is sincerely appreciated

dontlaugh commented:
This PR sets up all that.

It makes use of your new container, too, and runs 377 tests under xt/ with prove6.

finanalyst commented:
@dontlaugh ++ Sorry I missed the line where you draw in the rakuast-rakudoc-render image. My bad.
Merging now to see what happens. 
I can see two jobs, a test run and an integration run.
